New wheelhouse.txt parser can't handle comments.

## Checklist

 - [X] Confirmed this is an issue with charm-tools, not charmstore-client
 - [X ] Provide versions of tools used
 - [X ] Described the feature or ways to replicate the issue

## What version am I running?

I ran the following command: `snap info charm` and got the following ouput:

```
$ snap info charm
name:      charm
summary:   charm and charm-tools
publisher: Canonical✓
store-url: https://snapcraft.io/charm
contact:   https://discourse.juju.is/c/charming
license:   unset
description: |
  charmstore-client and charm-tools
commands:
  - charm
snap-id:      2Rryoc2ylScfbFl4eQtpntHD9iuZuMvt
tracking:     latest/stable
refresh-date: today at 08:48 BST
channels:
  latest/stable:    2.7.6 2020-07-16 (489) 117MB classic
  latest/candidate: 2.7.6 2020-07-16 (489) 117MB classic
  latest/beta:      2.7.6 2020-07-16 (489) 117MB classic
  latest/edge:      2.7.6 2020-07-16 (489) 117MB classic
installed:          2.7.6            (489) 117MB classic
```

I am using: (Your operating system)  Bionic

## Issue/Feature

Version of charm can't handle comments in wheelhouse.txt.  Previous version was just fine.

(with added debug):

```
parse_line: '# layer:openstack'
parse_line: '# charmhelpers.contrib.openstack.cert_utils charms with certificates relation'
parse_line: 'dnspython'
parse_line: '# charmhelpers.contrib.network all'
parse_line: 'netifaces>=0.10.4'
parse_line: '# charmhelpers.contrib.openstack.context all'
parse_line: '# psutil  # overridden by mysql-router'
parse_line: '# charms_openstack - pull from github until first release'
parse_line: '# git+https://github.com/openstack/charms.openstack.git#egg=charms.openstack  # overridden by mysql-router'
Traceback (most recent call last):
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/bin/charm-build", line 8, in <module>
    sys.exit(main())
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/builder.py", line 955, in main
    build()
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/builder.py", line 655, in __call__
    self.generate()
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/builder.py", line 599, in generate
    self.formulate_plan(layers)
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/builder.py", line 520, in formulate_plan
    self.plan = self.plan_layers(layers, output_files)
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/builder.py", line 372, in plan_layers
    output_files['wheelhouse.txt'].combine(existing_tactic)
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/charmtools/build/tactics.py", line 1072, in combine
    req = Requirement.parse(line)
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/requirements/requirement.py", line 221, in parse
    return cls.parse_line(line)
  File "/home/alex/Projects/Canonical/git/opendev.org/openstack/charm-mysql-router/.tox/build/lib/python3.6/site-packages/requirements/requirement.py", line 184, in parse_line
    assert local_match is not None, 'This should match everything'
AssertionError: This should match everything
```

It's a comment; it shouldn't even be being attempted to match.

Relevant code in `..lib/python3.6/site-packages/requirements/requirement.py` line 183:

```python
        elif '#egg=' in line:
            # Assume a local file match
            assert local_match is not None, 'This should match everything'
            groups = local_match.groupdict()
```

I'd argue that it shouldn't even get here if the line is a comment.

--- a/charmtools/build/tactics.py
+++ b/charmtools/build/tactics.py
@@ -8,7 +8,6 @@ from inspect import getargspec
 
 import requirements
 from path import Path as path
-from requirements.requirement import Requirement
 from ruamel import yaml
 from charmtools import utils
 from charmtools.build.errors import BuildError
@@ -1062,14 +1061,14 @@ class WheelhouseTactic(ExactMatch, Tactic):
         new_pkgs = set()
         for line in self.lines:
             try:
-                req = Requirement.parse(line)
+                req = next(requirements.parse(line))
                 new_pkgs.add(req.name)
-            except ValueError:
+            except (StopIteration, ValueError):
                 pass  # ignore comments, blank lines, etc
         existing_lines = []
         for line in existing.lines:
             try:
-                req = Requirement.parse(line)
+                req = next(requirements.parse(line))
                 # new explicit reqs will override existing ones
                 if req.name not in new_pkgs:
                     existing_lines.append(line)
@@ -1077,7 +1076,7 @@ class WheelhouseTactic(ExactMatch, Tactic):
                     existing_lines.append('# {}  # overridden by {}'
                                           ''.format(line,
                                                     self.layer.url))
-            except ValueError:
+            except (StopIteration, ValueError):
                 existing_lines.append(line)  # ignore comments, blank lines, &c
         self.lines = existing_lines + self.lines
         return self

--- a/tests/layers/whbase/wheelhouse.txt
+++ b/tests/layers/whbase/wheelhouse.txt
@@ -1,3 +1,4 @@
+# base-comment
 foo==1.0
 bar==1.0
 qux==1.0

--- a/tests/layers/whlayer/wheelhouse.txt
+++ b/tests/layers/whlayer/wheelhouse.txt
@@ -1,3 +1,4 @@
+# git+https://github.com/me/baz#egg=baz  # comment
 foo==2.0
 git+https://github.com/me/bar#egg=bar
 qux==2.0

--- a/tests/test_build.py
+++ b/tests/test_build.py
@@ -453,11 +453,13 @@ class TestBuild(unittest.TestCase):
                     self.assertEqual(len(Process._wheelhouses), 1)
                     self.assertEqual(Process._wheelhouses[0], [
                         '# layers/whbase',
+                        '# base-comment',
                         '# foo==1.0  # overridden by whlayer',
                         '# bar==1.0  # overridden by whlayer',
                         '# qux==1.0  # overridden by whlayer',
                         '',
                         '# whlayer',
+                        '# git+https://github.com/me/baz#egg=baz  # comment',
                         'foo==2.0',
                         'git+https://github.com/me/bar#egg=bar',
                         '# qux==2.0  # overridden by --wheelhouse-overrides',

